micro-rdk-nmea/src/parse_helpers/parsers.rs

```diff
@@ -0,0 +1,404 @@
+use bitvec::prelude::*;
+use std::{
+    marker::PhantomData,
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use micro_rdk::common::sensor::GenericReadingsResult;
+
+use super::{
+    enums::NmeaEnumeratedField,
+    errors::{NmeaParseError, NumberFieldError},
+};
+
+/// Cursor over a byte slice of data. The cursor can be moved by either consuming
+/// child-instances of `DataRead` or calling the `advance` function
+pub struct DataCursor<'a> {
+    data: &'a [u8],
+    bit_position: Rc<AtomicUsize>,
+}
+
+struct DataRead<'a> {
+    data: &'a [u8],
+    bit_position: Rc<AtomicUsize>,
+    bit_size: usize,
+}
+
+impl<'a> Drop for DataRead<'a> {
+    fn drop(&mut self) {
+        self.bit_position.fetch_add(self.bit_size, Ordering::SeqCst);
+    }
+}
+
+impl<'a> DataCursor<'a> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            data,
+            bit_position: Rc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    fn read(&self, bits: usize) -> Result<DataRead, NmeaParseError> {
```

Comment:
If you call `read` twice without dropping the `DataRead` object that resulted from the first `read` before the second call to `read`, what happens?

Reply:
It will be very bad. They could both read from the same starting place but potentially different ending places. Even if they do read to the same end position, when they're both dropped the bit position will advance twice as far as it should. Should I protect against this?

Reply:
Ideally, you would leverage the type system to make such a situation impossible. Just as a for instance, you could do something similar to how ScopeGuard works, where calling `read` on a `DataCursor` consumes the `DataCursor` and hands you a `DataRead` object that holds an inner `DataCursor`, which you must recover via `into_inner` or similar, which in turn consumes the `DataRead` and restores the `DataCursor`. That way, you go back and forth between them, but you can't end up with two `DataRead` objects obtained on the same `DataCursor` object. That might also let you get rid of the `Rc`? I'll bet @npmenard would have some even better ideas about how to structure it. This is just me thinking aloud.